### DISCUSSION TITLE
Build issue for UnrealEngine on Mac M1

### DISCUSSION THREAD

**Discussion by shkreza** [carla-simulator/carla#4848] [General]
I'm getting this linker error when building UnrealEngine for Carla on a Mac (using [these instructions](https://github.com/CarlaUnreal/UnrealEngine#mac)). I checked the MTLPP wrapper library which was built earlier in the process. The static lib file does have symbols for x86_64 architecture, which makes this error all more mysterious.

* Xcode version: 13.1
* Arch: Mac mini M1 (although build architecture below is x86_64)

**Build error output:**
```
Showing All Messages
Deploying ShaderCompileWorker Mac Development...
Deploying now!
Total execution time: 2.71 seconds
Running Engine/Binaries/DotNET/UnrealBuildTool.exe UE4Editor Mac Development  -architecture=x86_64
Showing All Messages
Building UE4Editor...
Performing 25 actions (8 in parallel)
Copying RadioEffectUnit.component to app bundle
[1/25] Link UE4Editor-MetalRHI.dylib
Undefined symbols for architecture x86_64:
  "typeinfo for mtlpp::CommandQueue", referenced from:
      typeinfo for ns::AutoReleased<mtlpp::CommandQueue> in Module.MetalRHI.1_of_3.cpp.o
  "typeinfo for mtlpp::ComputeCommandEncoder", referenced from:
      typeinfo for ns::AutoReleased<mtlpp::ComputeCommandEncoder> in Module.MetalRHI.1_of_3.cpp.o
  "typeinfo for mtlpp::ParallelRenderCommandEncoder", referenced from:
      typeinfo for ns::AutoReleased<mtlpp::ParallelRenderCommandEncoder> in Module.MetalRHI.1_of_3.cpp.o
  "typeinfo for mtlpp::CommandBuffer", referenced from:
      typeinfo for ns::AutoReleased<mtlpp::CommandBuffer> in Module.MetalRHI.1_of_3.cpp.o
      typeinfo for ns::AutoReleased<mtlpp::CommandBuffer> in Module.MetalRHI.2_of_3.cpp.o
  "typeinfo for mtlpp::Texture", referenced from:
      typeinfo for FMetalTexture in Module.MetalRHI.1_of_3.cpp.o
      typeinfo for ns::AutoReleased<mtlpp::Texture> in Module.MetalRHI.1_of_3.cpp.o
      typeinfo for FMetalTexture in Module.MetalRHI.2_of_3.cpp.o
      typeinfo for ns::AutoReleased<mtlpp::Texture> in Module.MetalRHI.2_of_3.cpp.o
      typeinfo for FMetalTexture in Module.MetalRHI.3_of_3.cpp.o
  "typeinfo for mtlpp::BlitCommandEncoder", referenced from:
      typeinfo for ns::AutoReleased<mtlpp::BlitCommandEncoder> in Module.MetalRHI.2_of_3.cpp.o
  "typeinfo for mtlpp::RenderCommandEncoder", referenced from:
      typeinfo for ns::AutoReleased<mtlpp::RenderCommandEncoder> in Module.MetalRHI.1_of_3.cpp.o
  "typeinfo for mtlpp::Device", referenced from:
      typeinfo for ns::AutoReleased<mtlpp::Device> in Module.MetalRHI.1_of_3.cpp.o
  "typeinfo for mtlpp::Buffer", referenced from:
      typeinfo for FMetalBuffer in Module.MetalRHI.1_of_3.cpp.o
      typeinfo for ns::AutoReleased<mtlpp::Buffer> in Module.MetalRHI.1_of_3.cpp.o
      typeinfo for ns::AutoReleased<mtlpp::Buffer> in Module.MetalRHI.2_of_3.cpp.o
      typeinfo for ns::AutoReleased<mtlpp::Buffer> in Module.MetalRHI.3_of_3.cpp.o
ld: symbol(s) not found for architecture x86_64
/Users/vsa/Code/UnrealEngine/clang:1:1: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by GustavoSilvera** [comment]
Hi, I've been having this same issue (M1 Max, XCode 12.2.1, x86 target) but I found that I could at least bypass this error by switching to this commit [c3caf7b6bf12ae4c8e09b606f10a09776b4d1f38](https://github.com/CarlaUnreal/UnrealEngine/commit/c3caf7b6bf12ae4c8e09b606f10a09776b4d1f38) and making all other critical changes myself. Interestingly the carla changes did not seem like they should cause problems but apparently they do.

Then I could build Carla on my M1 and it runs like a champ

**Reply by chchch981** [reply]
can I ask how you can build Carla on Mac m1?

**Reply by GustavoSilvera** [reply]
Sure! I actually just made a quick pr to Carla that you can find [here](https://github.com/carla-simulator/carla/pull/5086). For specific build instructions (though I am happy to answer more questions) see the [Docs/build_mac.md](https://github.com/GustavoSilvera/carla-mac/blob/m1/Docs/build_mac.md) file that I created.

**Reply by chchch981** [reply]
Thanks! I follow your directions for Mac  
but I have question for your comment. 

make sure the ARCH (for IS_MAC) is set to "x86_64" 
what is meaning?  
does it meaning for on rosetta?  

but in my roestta terminal, 
(carla) beckk@choebaegjin-ui-MacBookAir carla % make LibCarla 
clang 8 is required, but it's not installed.
make: *** [setup] Error 1

have problem like this.

**Reply by GustavoSilvera** [reply]
Ok so two points:
Your first point is that if you go to [Util/Buildtools/Environment.sh:19](https://github.com/GustavoSilvera/carla-mac/blob/14a8f993fb6ade6dc5db5e0edaec03f6a6464b89/Util/BuildTools/Environment.sh#L19) then you'll want to use `ARCH="x86_64"` for running the UE4 game on rosetta. However, if you want to build the PythonAPI/check, then you'll need to change that to `"arm64"`

To your second point, that error should no longer happen since [the setup](https://github.com/GustavoSilvera/carla-mac/blob/14a8f993fb6ade6dc5db5e0edaec03f6a6464b89/Util/BuildTools/Setup.sh#L41) only requires `clang` not `clang-8`. Make sure you're on the [right repo and right branch (`m1`)](https://github.com/GustavoSilvera/carla-mac/tree/m1)

**Reply by chchch981** [reply]
thanks I solve that problem but I have another issue.

when I build like this : 
# make sure the ARCH (for IS_MAC) is set to "arm64"
mv Build Build.x86 # denoting the current (x86) build as such

make LibCarla # now in arm64!
...

make PythonAPI # now in arm64
...

make check # unit tests to verify everything works!

make PythonAPI has some problem 
BuildPythonAPI.sh: Building Python API for Python 3.
Error processing line 1 of /Users/beckk/miniforge3/envs/carla/lib/python3.8/site-packages/distutils-precedence.pth:

  Traceback (most recent call last):
    File "/Users/beckk/miniforge3/envs/carla/lib/python3.8/site.py", line 169, in addpackage
      exec(line)
    File "<string>", line 1, in <module>
  ModuleNotFoundError: No module named '_distutils_hack'

Remainder of file ignored
compiling:
  - source/libcarla/libcarla.cpp
running bdist_egg
running egg_info
writing source/carla.egg-info/PKG-INFO
writing dependency_links to source/carla.egg-info/dependency_links.txt
writing top-level names to source/carla.egg-info/top_level.txt
reading manifest file 'source/carla.egg-info/SOURCES.txt'
writing manifest file 'source/carla.egg-info/SOURCES.txt'
installing library code to build/bdist.macosx-11.0-arm64/egg
running install_lib
running build_py
running build_ext
building 'carla.libcarla' extension
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/Users/beckk/miniforge3/envs/carla/include -arch arm64 -I/Users/beckk/miniforge3/envs/carla/include -arch arm64 -Idependencies/include -I/Users/beckk/miniforge3/envs/carla/include/python3.8 -c source/libcarla/libcarla.cpp -o build/temp.macosx-11.0-arm64-3.8/source/libcarla/libcarla.o -isystem dependencies/include/system -fPIC -std=c++14 -Werror -Wall -Wextra -Wpedantic -Wno-self-assign-overloaded -Wdeprecated -Wno-shadow -Wuninitialized -Wunreachable-code -Wpessimizing-move -Wold-style-cast -Wnull-dereference -Wduplicate-enum -Wnon-virtual-dtor -Wheader-hygiene -Wconversion -Wfloat-overflow-conversion -DBOOST_ERROR_CODE_HEADER_ONLY -DLIBCARLA_WITH_PYTHON_SUPPORT -stdlib=libc++ -DLIBCARLA_IMAGE_WITH_PNG_SUPPORT=true
In file included from source/libcarla/libcarla.cpp:7:
In file included from dependencies/include/carla/Memory.h:9:
In file included from dependencies/include/system/boost/enable_shared_from_this.hpp:16:
In file included from dependencies/include/system/boost/smart_ptr/enable_shared_from_this.hpp:16:
In file included from dependencies/include/system/boost/smart_ptr/weak_ptr.hpp:16:
In file included from /Users/beckk/miniforge3/envs/carla/bin/../include/c++/v1/memory:670:
In file included from /Users/beckk/miniforge3/envs/carla/bin/../include/c++/v1/typeinfo:61:
In file included from /Users/beckk/miniforge3/envs/carla/bin/../include/c++/v1/exception:81:
In file included from /Users/beckk/miniforge3/envs/carla/bin/../include/c++/v1/__memory/base.h:14:
In file included from /Users/beckk/miniforge3/envs/carla/bin/../include/c++/v1/__debug:14:
In file included from /Users/beckk/miniforge3/envs/carla/bin/../include/c++/v1/iosfwd:95:
/Users/beckk/miniforge3/envs/carla/bin/../include/c++/v1/wchar.h:119:15: fatal error: 'wchar.h' file not found
#include_next <wchar.h>
              ^~~~~~~~~
1 error generated.
error: command 'clang' failed with exit status 1
 
here is my terminal  how can I solve this..?

**Reply by GustavoSilvera** [reply]
Make sure XCode is installed. Run `xcode-select --install` (Check out [this error](https://stackoverflow.com/questions/26185978/macos-wchar-h-file-not-found))

**Reply by chchch981** [reply]
oh,, sorry before make PythonAPI 
there is one more problem
when I do "make launch",  BuildCarlaUE4.sh: Build CarlaUE4 project.
Command line invocation:
    /Applications/Xcode.app/Contents/Developer/usr/bin/xcodebuild -scheme CarlaUE4 -target CarlaUE4Editor -UseModernBuildSystem=YES

User defaults from command line:
    IDEPackageSupportUseBuiltinSCM = YES
    UseModernBuildSystem = YES

xcodebuild: error: The directory /Users/beckk/carla-mac/Unreal/CarlaUE4 does not contain an Xcode project, workspace or package.

make: *** [launch] Error 66

there is such problem.


But I installed UE4, by installing epic game launcher. UE4 version is 4.26.2.  NOT BY GitHub.
if I have wrong UE4, could you tell me the right way to install UE4?

Where did it go wrong?  Im very sorry for bothering,,


-----------------I reinstall UE4  by https://github.com/CarlaUnreal/UnrealEngine/tree/c3caf7b6bf12ae4c8e09b606f10a09776b4d1f38---------------


setup.command and GenerateProjectFiles.command are OK

but when I build UE4.xcworkspace , I have build failed

Prepare build 
: The Legacy Build System will be removed in a future release. You can configure the selected build system and this deprecation message in File > Workspace Settings. --> I solve this my Xcode - file - working space settings - all section to new

**Reply by GustavoSilvera** [reply]
Right, you're almost done then. Once you run the setup and GenerateProjectFiles (either through `sh` or the `.command`) then you should be able to `make launch` and have everything build properly (the `UseModernBuildSystem=YES` is for the Legacy Build System error... You will probably need to make some edits to the Engine itself following Carla's commits in order for everything to compile (since their `carla` branch fails due to the error of this original discussion). But running the `BuildCarlaUE4.sh` script should work to generate the `CarlaUE4.xcworkspace` which then gets used in `xcodebuild`.

**Reply by chchch981** [reply]
./Setup.sh
Running bundled mono, version: Mono JIT compiler version 5.16.0.220 (2018-06/bb3ae37d71a Fri Nov 16 17:12:11 EST 2018)
ln: Engine/Binaries/ThirdParty/Mono/Mac/lib/libmsvcrt.dylib: File exists
(carla) beckk@choebaegjin-ui-MacBookAir UnrealEngine-4.62 % ./GenerateProjectFiles.sh

Setting up Unreal Engine 4 project files...

Running bundled mono, version: Mono JIT compiler version 5.16.0.220 (2018-06/bb3ae37d71a Fri Nov 16 17:12:11 EST 2018)
2022-01-20 11:48:35.051 defaults[10523:107702] 
The domain/default pair of (com.epicgames.ue4, MonoAOT) does not exist
Generating data for project indexing... 100%
Writing project files...    
Getting signing information via /Users/beckk/UnrealEngine-4.62/Engine/Build/BatchFiles/Mac/RunMono.sh "/Users/beckk/UnrealEngine-4.62/Engine/Binaries/DotNET/IOS/IPhonePackager.exe" signing_match Engine -bundlename com.YourCompany.UE4Game
Getting signing information via /Users/beckk/UnrealEngine-4.62/Engine/Build/BatchFiles/Mac/RunMono.sh "/Users/beckk/UnrealEngine-4.62/Engine/Binaries/DotNET/IOS/IPhonePackager.exe" signing_match Engine -bundlename com.YourCompany.UE4Game -tvos
Writing project files... 100%

and  for env variable 
export UE4_ROOT=/Users/beckk/UnrealEngine-4.62

and  "make launch"

now I m on building this  774/1766

is there upcoming issue in my routine?

**Reply by GustavoSilvera** [reply]
That seems fine, as long as the `CarlaUE4.xcworkspace` has been created in the `Unreal/CarlaUE4` directory

**Reply by chchch981** [reply]
[1437/1766] Copy libembree.2.14.0.dylib

MkDir /Users/beckk/UnrealEngine-4.62/Engine/Binaries/Mac/UE4Editor.app (in target 'CarlaUE4' from project 'CarlaUE4')
    cd /Users/beckk/carla-mac/Unreal/CarlaUE4/Intermediate/ProjectFiles
    /bin/mkdir -p /Users/beckk/UnrealEngine-4.62/Engine/Binaries/Mac/UE4Editor.app

warning: Skipping code signing because the target does not have an Info.plist file and one is not being generated automatically. (in target 'CarlaUE4' from project 'CarlaUE4')
** BUILD FAILED **


The following build commands failed:
	ExternalBuildToolExecution CarlaUE4_Build (in target 'CarlaUE4_Build' from project 'CarlaUE4')
(1 failure)
make: *** [launch] Error 65


in my file 

(carla) beckk@choebaegjin-ui-MacBookAir carla-mac % ls ./Unreal/CarlaUE4/
Binaries		CarlaUE4.xcworkspace	Intermediate
CarlaUE4.png		Config			Plugins
CarlaUE4.uproject	Content			Source


could you give me hand?

I'm very sorry,,

**Reply by GustavoSilvera** [reply]
Hi, no worries. I simplified some of the build process for Unreal in my latest commit to the `m1` branch. Take a look [here](https://github.com/GustavoSilvera/carla-mac/tree/m1). Notably, you can now build Unreal Engine normally (updated docs in [`build_mac.md`](https://github.com/GustavoSilvera/carla-mac/blob/m1/Docs/build_mac.md)).

**Reply by GustavoSilvera** [reply]
Update: See [PR here](https://github.com/CarlaUnreal/UnrealEngine/pull/3)

**Reply by itserror404** [reply]
hey were you able to solve this issue? if yes, can you share please?

**Reply by svenwinkelmann** [reply]
I solved the infoplist error with: xcodebuild -scheme UE4 -target UE4 -UseModernBuildSystem=YES GENERATE_INFOPLIST_FILE=YES

**Comment by GustavoSilvera** [comment]
Hi all, so I've managed to locate the bug and create a solution. Here is the [PR](https://github.com/CarlaUnreal/UnrealEngine/pull/3) to the CarlaUnreal fork. For now, I recommend you use [this fork](https://github.com/GustavoSilvera/UnrealEngine) and switch to the `mac-dev` branch.

**Reply by ghost** [reply]
Hello Gustavo, it seems like the links aren't working anymore. Could you please update them?

**Comment by chchch981** [comment]
Hello

Thanks for your UE4 building 

 Ive don UE4 build for your helping 

But I have some problem in  “make launch” in rosetta x86




My finder said “ can’t open the application

What’s problem with me ?

My Mac is Mac book air m1 , 8ram, 256 g

But I check 130 + GB my storage

**Reply by GustavoSilvera** [reply]
When you run `make launch` the UE4.26 executable should launch on its own. But when you run `make package` it'll create a `.app` which is essentially a directory, so you'll need to open it up with `open` in the terminal.

**Comment by chchch981** [comment]
Hello there

I change my laptop for Carla lol to MacBook 14 m1pro

Anyway I have problem  in
brew install --build-from-source mono   

 
==> Installing dependencies for mono: cmake, pkg-config, gdbm, mpdecimal, ca-certificates, ***@***.***, readline, sqlite, xz and ***@***.***
==> Installing mono dependency: cmake
==> Pouring cmake--3.22.1.arm64_monterey.bottle.tar.gz
🍺  /opt/homebrew/Cellar/cmake/3.22.1: 2,986 files, 41.6MB
==> Installing mono dependency: pkg-config
==> Pouring pkg-config--0.29.2_3.arm64_monterey.bottle.tar.gz
🍺  /opt/homebrew/Cellar/pkg-config/0.29.2_3: 11 files, 676.4KB
==> Installing mono dependency: gdbm
==> Pouring gdbm--1.22.arm64_monterey.bottle.tar.gz
🍺  /opt/homebrew/Cellar/gdbm/1.22: 24 files, 1MB
==> Installing mono dependency: mpdecimal
==> Pouring mpdecimal--2.5.1.arm64_monterey.bottle.tar.gz
🍺  /opt/homebrew/Cellar/mpdecimal/2.5.1: 71 files, 2.2MB
==> Installing mono dependency: ca-certificates
==> Pouring ca-certificates--2021-10-26.all.bottle.tar.gz
==> Regenerating CA certificate bundle from keychain, this may take a while...
🍺  /opt/homebrew/Cellar/ca-certificates/2021-10-26: 3 files, 208.5KB
==> Installing mono dependency: ***@***.***
==> Pouring ***@***.***_monterey.bottle.tar.gz
🍺  ***@***.***/1.1.1m: 8,081 files, 18MB
==> Installing mono dependency: readline
==> Pouring readline--8.1.2.arm64_monterey.bottle.tar.gz
🍺  /opt/homebrew/Cellar/readline/8.1.2: 48 files, 1.7MB
==> Installing mono dependency: sqlite
==> Pouring sqlite--3.37.2.arm64_monterey.bottle.tar.gz
🍺  /opt/homebrew/Cellar/sqlite/3.37.2: 11 files, 4.3MB
==> Installing mono dependency: xz
==> Pouring xz--5.2.5.arm64_monterey.bottle.tar.gz
🍺  /opt/homebrew/Cellar/xz/5.2.5: 95 files, 1.4MB
==> Installing mono dependency: ***@***.***
==> Pouring ***@***.***_monterey.bottle.tar.gz
==> ***@***.***/3.10.1/bin/python3 -m ensurepip
==> ***@***.***/3.10.1/bin/python3 -m pip install -v --no-d
🍺  ***@***.***/3.10.1: 3,134 files, 57.9MB
==> Installing mono
==> Patching
==> Applying 3070886a1c5e3e3026d1077e36e67bd5310e0faa.patch
patching file packaging/MacSDK/fsharp-layout.sh
==> Applying configure-big_sur.diff
patching file configure
Hunk #1 succeeded at 10928 (offset 2195 lines).
==> ./configure --prefix=/opt/homebrew/Cellar/mono/6.12.0.122_1 --enable-nls=no
==> make
Last 15 lines from /Users/choebaegjin/Library/Logs/Homebrew/mono/02.make:
6 errors generated.
make[5]: *** [libmonoutils_la-mono-context.lo] Error 1
make[5]: *** Waiting for unfinished jobs....
mv -f .deps/libmonoutils_la-mono-property-hash.Tpo .deps/libmonoutils_la-mono-property-hash.Plo
mv -f .deps/libmonoutils_la-mono-value-hash.Tpo .deps/libmonoutils_la-mono-value-hash.Plo
mv -f .deps/libmonoutils_la-lock-free-queue.Tpo .deps/libmonoutils_la-lock-free-queue.Plo
mv -f .deps/libmonoutils_la-lock-free-array-queue.Tpo .deps/libmonoutils_la-lock-free-array-queue.Plo
mv -f .deps/libmonoutils_la-hazard-pointer.Tpo .deps/libmonoutils_la-hazard-pointer.Plo
mv -f .deps/libmonoutils_la-lock-free-alloc.Tpo .deps/libmonoutils_la-lock-free-alloc.Plo
mv -f .deps/libmonoutils_la-mono-error.Tpo .deps/libmonoutils_la-mono-error.Plo
make[4]: *** [all-recursive] Error 1
make[3]: *** [all] Error 2
make[2]: *** [all-recursive] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
Why I have this issue … ?

**Reply by GustavoSilvera** [reply]
Not sure. Also unsure if you need mono installed manually or if UE4 brings a version of mono packaged with it. But in any case, try looking at this [build documentation](https://www.mono-project.com/docs/compiling-mono/mac/).

**Comment by chchch981** [comment]
Okay I skipped this process , I think UE4 brings a version of mono !!

I think almost done lastly I have a question.

I run 'make package' in rosetta. But I think it has very long time to compile 

Does this usually take a long time?

How long did it take for you?

**Reply by GustavoSilvera** [reply]
Yes, make package takes a while (and sometimes hangs, in which case just press enter on the keyboard). For me it took ~1h the first time and quicker afterwards (once most things were cooked and cached).

**Reply by chchch981** [reply]
hi I think I install almost done ! 

but I have different egg.  not carla-0.9.13-py3.8-macosx-11.0-arm64.egg but in my, carla-0.9.13-py3.8-macosx-10.9-x86_64.egg

Perhaps that's why there was an error.

every pythonAPI has 

ImportError: dlopen(/Users/choebaegjin/Library/Caches/Python-Eggs/carla-0.9.13-py3.8-macosx-10.9-x86_64.egg-tmp/carla/libcarla.cpython-38-darwin.so, 0x0002): symbol not found in flat namespace '_png_create_info_struct' 

I did make launch in x86, and then changed arm64.
but my egg file is till x86_64  
Can this be a problem?

**Reply by GustavoSilvera** [reply]
Hm yeah, this is where it gets a bit tricky but what worked for me is having the `make launch` in x86 mode. This has the `make LibCarla` dependency which all gets placed into `Build/` then I rename that folder to `Build.x86` and switch the mode to `arm64` in `Environment.sh` and run `make PythonAPI` here. This will create a **new** `Build/` folder and should make the python egg be an `arm64.egg`. Then note that whenever you `make launch` again you'll need the old(x86) `Build/` folder which you can do by renaming the new one to `Build.arm64` and renaming the (old) `Build.x86` back to `Build/`

Hope this helps!

**Reply by chchch981** [reply]
hum I tried follow 

dir : carla-mac  -> make LibCarla in rosetta terminal
dir : carla-mac -> make launch in rosetta termina

dir : Carla-mac/Utils/BuildTools/Enviroment.sh  line 19 : x86_64 ->. arm64

dir : carla-mac  -> make LibCarla in terminal
dir : carla-mac  -> make PythonAPI in terminal

but still x86_64.egg in dir : carla-mac/PythonAPI/dist

is there any wrong ?

I'm sorry, but don't you want to film it?
Like youtube

**Reply by GustavoSilvera** [reply]
Did you make sure to rename the old `Build/` directory (from the `x86` build) to `Build.x86` and **then** do the arm64 build? then once the arm64 .egg is created you can rename it back

**Comment by chchch981** [comment]
Yep, but It is still x86_64 
I don’t know what is wrong lol..

**Reply by GustavoSilvera** [reply]
That it weird because I tried it again and it worked for me. Ok maybe try this: restart the Carla repo by cloning a fresh copy and FIRST compiling for arm64 (`make PythonAPI`) (when the environment is set to arm64 mode) then rename that `Build` folder to `Build.arm64` and set the environment back to `x86_64` then run `make LibCarla` (NOT `make PythonAPI`) and `make launch/package` should work as well.

Note this should only affect the .egg in Carla/PythonAPI/Carla/dist/` in the main for, not for a packaged build (you'll need to copy it over)

**Comment by chchch981** [comment]
## you can see my capture in your mail!


Hmm.. I think there’s no wrong point with me..

If I want init all process where should I start now? 

For installing UE4?

**Reply by GustavoSilvera** [reply]
No, once UE4 is working you're good there. The problem is with the PythonAPI apparently so you should just restart the Carla build as described above. (First do PythonAPI in arm64 mode then switch to x86 and do the normal build (without PythonAPI)).

**Reply by chchch981** [reply]
I solved this problem. ! 
my Conda has the problem  I had installed Conda for x86 so that's why.

https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-MacOSX-arm64.sh 
I install this Conda so  I can find arm64.egg !

**Comment by lokesh-c-das** [comment]
Hi 
I am facing this error while importing carla. I did everything in correct and I have also installed Miniforge3-MaxOSX-arm64.sh

dlopen(/Users/ldas/Library/Caches/Python-Eggs/carla-0.9.13-py3.8-macosx-11.0-arm64.egg-tmp/carla/libcarla.cpython-38-darwin.so, 0x0002): symbol not found in flat namespace '_png_create_info_struct'

Any suggestions would be very helpful.

Thanks

**Comment by yhamadi75** [comment]
Hello!
This is a pretty old thread, but I want to install Carla on a Mac M1, and no simple way to do that. The links provided in the discussion above are dead, anyone who went through the same issue is welcome to help!
Also considering going through a windows or linux virtual machine if that's easier!
Thanks,

**Reply by GustavoSilvera** [reply]
Going through a Linux VM would certainly be easier. But the links in this discussion are mostly all alive? (I haven't found a dead one)

Have you made sure to join the Epic Games organization on GitHub to get access to their github repo/fork links?

**Reply by newgamerboy** [reply]
Hi,Gustavo.I  tried  to build ue4 on arm 64,but when i executed setup.sh, an error occurred: ./Setup.sh
Fixing inconsistent case in filenames.
Setting up Mono

Unhandled Exception:
System.TypeInitializationException: The type initializer for 'GitDependencies.Program' threw an exception. ---> System.TypeInitializationException: The type initializer for 'Sys' threw an exception. ---> System.DllNotFoundException: System.Native assembly:<unknown assembly> type:<unknown type> member:(null)
  at (wrapper managed-to-native) Interop+Sys.LChflagsCanSetHiddenFlag()
  at Interop+Sys..cctor () [0x00000] in <d636f104d58046fd9b195699bcb1a744>:0
   --- End of inner exception stack trace ---
  at Interop.GetRandomBytes (System.Byte* buffer, System.Int32 length) [0x00000] in <d636f104d58046fd9b195699bcb1a744>:0
  at System.Guid.NewGuid () [0x00000] in <d636f104d58046fd9b195699bcb1a744>:0
  at GitDependencies.Program..cctor () [0x00000] in <0916da07a3c04a6aa2e0f66d38c790da>:0
   --- End of inner exception stack trace ---
[ERROR] FATAL UNHANDLED EXCEPTION: System.TypeInitializationException: The type initializer for 'GitDependencies.Program' threw an exception. ---> System.TypeInitializationException: The type initializer for 'Sys' threw an exception. ---> System.DllNotFoundException: System.Native assembly:<unknown assembly> type:<unknown type> member:(null)
  at (wrapper managed-to-native) Interop+Sys.LChflagsCanSetHiddenFlag()
  at Interop+Sys..cctor () [0x00000] in <d636f104d58046fd9b195699bcb1a744>:0
   --- End of inner exception stack trace ---
  at Interop.GetRandomBytes (System.Byte* buffer, System.Int32 length) [0x00000] in <d636f104d58046fd9b195699bcb1a744>:0
  at System.Guid.NewGuid () [0x00000] in <d636f104d58046fd9b195699bcb1a744>:0
  at GitDependencies.Program..cctor () [0x00000] in <0916da07a3c04a6aa2e0f66d38c790da>:0
   --- End of inner exception stack trace ---
Result: 1
This seems to be a missing System.Native library, but I don't know how to resolve it.

**Reply by skylinestars** [reply]
@newgamerboy I encountered the same problem as you, have you solved it

**Comment by yhamadi75** [comment]
Thanks for the info Gustavo, I have joined the Epic org, and should be able to try an M1 built!
Cheers,

**Comment by svenwinkelmann** [comment]
Thanks @GustavoSilvera for your great work! With your steps I was able to get Unreal and Carla running on my M1 Max, Ventura 13.1, XCode 14.2.

Some remarks I stumbled across:
- Unreal build: "Unable to instantiate module 'WebBrowser': System.IO.DirectoryNotFoundException" => create the mentioned folder UnrealCEFSubProcess.app manually
- Duplicate HTTP symbols you get rid of with changes in [libwebsockets.h](https://github.com/EpicGames/UnrealEngine/commit/2cd6352a6928d0ba38138a72d1aa8d1fdd882a68#diff-cf5e4dfefd794af1b69729cf4ac93d895d7559b6c3312a02a7a005118d3eee3a)
- Carla: I used the changes in this [commit](https://github.com/GustavoSilvera/carla/commit/20fb4586da0d1583267717d43bbbf7421aca4ce7)
- Pyenv did not work out, with conda it's fine
- I had huge problems with boost-python, because I had installed a global version with brew. Errors where [Symbol not found / Expected in: flat namespace](https://stackoverflow.com/questions/35006614/what-does-symbol-not-found-expected-in-flat-namespace-actually-mean) and "Error:  540: support for exception handling is disabled; use --exceptions to enable try {" ==> the only solution I have found was to deinstall boost with brew and use the one which is shipped with carla.

I hope these comments help someone. Carla runs like a charm on my M1 :)

**Reply by pixeloverflow** [reply]
solution magic working for me

**Comment by SevillaFe** [comment]
Hello @GustavoSilvera first of all, thank you for your work!! 

I'm trying out to follow the steeps from here [Mac build](https://github.com/JesusAnaya/carla/blob/macosx-dev/Docs/build_mac.md) but I cannot find your GitHub fork described in this steep git clone https://github.com/GustavoSilvera/UnrealEngine --depth 1 -b mac-dev

did you delete it? or did you rename it? could you help me please.

Thank you

**Reply by SevillaFe** [reply]
Edit:
I suppose that I have to wait at least 24 hours to get access and to see EpicGames/UnrealEngine repo after I linked my Github account to Epic, isn't?

**Reply by SevillaFe** [reply]
Hello again,
it worked like a charm, just waiting a few minutes I got access.

then I can continue with the build .

**Comment by VitalkaaNX** [comment]
Evre one!

**Comment by ilyLunaa** [comment]
This does not work as the output gives me a "remote server returned an error: (403) Forbidden. (WebException)" error while executing the setup.sh file from the fork.  I am logged in, everything is correct aswell.

**Comment by newgamerboy** [comment]
No.I have cancelled this project.



---Original---
From: ***@***.***&gt;
Date: Mon, Aug 19, 2024 16:38 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [carla-simulator/carla] Build issue for UnrealEngine on Mac M1(Discussion #4848)




 
@newgamerboy I encountered the same problem as you, have you solved it
